### PR TITLE
Can retrieve Annotations from DataObject properties

### PR DESCRIPTION
**Pull request by Cafeinoman**
Add the possibility to use custom annotation in dataobject generation. Annotation members can also be retrieved.
See issue #115

### CODE REVIEW

File: src/main/java/io/vertx/codegen/type/AnnotationValueTypeInfo.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
+package io.vertx.codegen.type;
+
+import java.util.List;
+
+/**
+ * @author <a href="mailto:cafeinoman@openaliasbox.org>Francois Delalleau</a>
+ */
+public class AnnotationValueTypeInfo {
```
Comment (vietj) [inline src/main/java/io/vertx/codegen/type/AnnotationValueTypeInfo.java]: should be named `AnnotationMemberInfo`

Comment (vietj) [inline src/main/java/io/vertx/codegen/type/AnnotationValueTypeInfo.java]: I'm not sure why this class is needed and why instead we don't simply use a Map<String, Object> to represents the value of an annotation.

Comment (Cafeinoman) [inline src/main/java/io/vertx/codegen/type/AnnotationValueTypeInfo.java]: Yes we could, but the type cast must be done after value retrieving...

Comment (vietj) [inline src/main/java/io/vertx/codegen/type/AnnotationValueTypeInfo.java]: well type cast must be done, except that calling getShortValue() does the cast automatically but it still happen. I prefer to use a Map to limit the API surface and remain clear in the intent.

Comment (Cafeinoman) [inline src/main/java/io/vertx/codegen/type/AnnotationValueTypeInfo.java]: ok, I make the change

Comment (Cafeinoman) [inline src/main/java/io/vertx/codegen/type/AnnotationValueTypeInfo.java]: done

**Comment by Cafeinoman** [comment]
I'm going to extend the capacity to retrieve annotations to ClassModel too...

**Comment by Cafeinoman** [comment]
Hi,
any news on this?

**Comment by Cafeinoman** [comment]
@vietj is the idea abandoned?

File: src/main/java/io/vertx/codegen/type/AnnotationTypeInfo.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author <a href="mailto:cafeinoman@openaliasbox.org>Francois Delalleau</a>
+ */
+public class AnnotationTypeInfo {
```
Comment (vietj) [inline src/main/java/io/vertx/codegen/type/AnnotationTypeInfo.java]: this should be named differently as this is not an annotation type but rather an annotation value, so something like `AnnotationValueInfo` seems better

File: src/test/java/io/vertx/test/codegen/testdataobject/Annotated.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import io.vertx.test.codegen.annotations.LongValuedAnnotation;
+import io.vertx.test.codegen.annotations.ShortValueAnnotation;
+import io.vertx.test.codegen.annotations.StringValuedAnnotation;
+import io.vertx.test.codegen.annotations.TestEnum;
+
+/**
+ * @author <a href="mailto:cafeinoman@openaliasbox.org>Francois Delalleau</a>
+ */
+@DataObject
+public class Annotated {
```
Comment (vietj) [inline src/test/java/io/vertx/test/codegen/testdataobject/Annotated.java]: rename to `AnnotatedDataObject`

File: src/main/java/io/vertx/codegen/ClassModel.java [hunk 4]
```diff
@@ -244,6 +258,14 @@ public TypeInfo getHandlerSuperType() {
     return null;
   }
 
+  public List<AnnotationTypeInfo> getAnnotations() {
```
Comment (vietj) [inline src/main/java/io/vertx/codegen/ClassModel.java]: missing doc

File: src/main/java/io/vertx/codegen/ClassModel.java [hunk 5]
```diff
@@ -244,6 +258,14 @@ public TypeInfo getHandlerSuperType() {
...
+    return annotations;
+  }
+
+  public Map<String, List<AnnotationTypeInfo>> getMethodAnnotationsMap() {
```
Comment (vietj) [inline src/main/java/io/vertx/codegen/ClassModel.java]: simply `methodAnnotations`

File: src/main/java/io/vertx/codegen/type/AnnotationTypeInfo.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+  public String getSimpleName() {
+    return simpleName;
+  }
+
+  public String getPackageName() {
+    return packageName;
+  }
+
+  public List<String> getMembersNames() {
```
Comment (vietj) [inline src/main/java/io/vertx/codegen/type/AnnotationTypeInfo.java]: why not keep the `Set<String>`

File: src/main/java/io/vertx/codegen/type/AnnotationTypeInfo.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    return new ArrayList<>(values.keySet());
+  }
+
+  public Object getMember(String name) {
```
Comment (vietj) [inline src/main/java/io/vertx/codegen/type/AnnotationTypeInfo.java]: what's wrong with `get(name)` ?

Comment (Cafeinoman) [inline src/main/java/io/vertx/codegen/type/AnnotationTypeInfo.java]: Thought it was clearer and in line with getMembersNames()...

File: src/main/java/io/vertx/codegen/ClassModel.java (hunk 4 above)
Comment (Cafeinoman) [inline src/main/java/io/vertx/codegen/ClassModel.java]: Javadoc or doc in package-info.java? Or both...

Comment (vietj) [inline src/main/java/io/vertx/codegen/ClassModel.java]: simple javadoc

Comment (vietj) [inline src/main/java/io/vertx/codegen/ClassModel.java]: also a few words should be in the global doc

**Comment by Cafeinoman** [comment]
All done, except for AnnotationValueInfo#getMember

File: src/main/java/io/vertx/codegen/type/AnnotationTypeInfoFactory.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import javax.lang.model.util.Types;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * @author <a href="mailto:cafeinoman@openaliasbox.org>Francois Delalleau</a>
+ */
+public class AnnotationTypeInfoFactory {
```
Comment (vietj) [inline src/main/java/io/vertx/codegen/type/AnnotationTypeInfoFactory.java]: I think it should be renamed as well to `AnnotationValueFactory `

**Pull request merged**